include/deal.II/base/vectorization.h

```diff
@@ -656,6 +656,16 @@ class VectorizedArray
     base_ptr[offsets[0]] = data;
   }
 
+  /**
+   * Returns horizontal sum of data field.
+   */
+  DEAL_II_ALWAYS_INLINE
+  Number
+  horizontal_add()
```

Comment:
I'm not familiar with the term "horizontal sum" or "horizontal add", and I suspect that others might not either. Would you mind adding a definition, perhaps in a formula, what this operation does to the elements of the array?

Reply:
The name comes from Intel, say this one: https://www.felixcloutier.com/x86/phaddw:phaddd
Regarding the name, I agree that this is not a very intuitive name. I do not like the `std::accumulate` name too much, either, because to me accumulate would be across some iterator range starting with some initial entry, where it is here the sum **within** a vector. I would prefer more verbose name, `sum_within_vector` or `accumulate_within_vector`. Another question without having checked myself: Does the SVE (scalable vector extensions of ARM) have a similar operation, and what name are they using? They tend to have good names, so it might worth checking.

Reply:
As @kronbichler says, the name comes from Intel and their `...hadd...` intrinsics. They call what our `operator+=` implements a _vertical_ addition (where each lane of a `VectorizedArray` does the addition which results in new `VectorizedArray` with same width), and the summation over the lanes (with the underlying floating type as a result) a _horizontal_ addition.

SVE has an instruction called _ADDV: Tree-based floating-point addition reduction_ which comes close to what happens.

I could live with the names `horizontal_add/horizontal_sum` or `accumulate` but we can also use a more verbose name here. (And/or a comment)

Reply:
If going by SVE, we would pick 'addition_reduction', which is also an option. I prefer `add_within_vector`, but I am fine with both `horizontal_add` and `accumulate...` as well.

Reply:
Why not just `vec.sum()`?

But to be clear, I don't actually care about the name as long as the documentation has a formula that explains what the function does :-)

Reply:
As discussed today, let's go with `sum()`. That is essentially what happens. Any further suggestions for documentation? Should we mention that we use a tree-like reduction?

Reply:
Tree-like reduction is good to mention, yes.

Reply:
I'll defer to @kronbichler in this case. But as a general rule, documentation should state *what* a function does, not *how* it does it. The latter is subject to change, and is not of interest to the reader of the documentation anyway. The *how* can be documented in in-comment commentary, but it should generally not be part of the documentation.